File: c++/leetcode/array/Median_of_Two_Sorted_Arrays.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main(int argc, char *argv[]) {
    Solution s;
    vector<int> nums1 = new vector<int>();
    vector<int> nums2 = new vector<int>();

    for(int i = 1;i<10;i++){
        nums1.push_back(i);
    }
    for(int i= 1;i<10;i=i+2){
        nums2.push_back(i);
    }

    s.findMedianSortedArrays(nums1,nums2);
    return 0;
}

class Solution {
public:
    double findMedianSortedArrays(vector<int> &nums1, vector<int> &nums2) {
        int k = nums1.size() + nums2.size();
        if(k%2 == 0){
            return min(getKthElement(nums1, nums2, k/2+1),getKthElement(nums1, nums2,k/2))/2.0;
        }else{
            return getKthElement(nums1,nums2,k/2);
        }
    }

    double getKthElement(vector<int> &nums1, vector<int> &nums2,int k){
        int index1 = 0;
        int index2 = 0;

        int m = nums1.size();
        int n = nums2.size();

        while (true){
            if (index1 == m){
                return nums2[index2+k-1];
            }
            if(index2 == n){
                return nums1[index1 +k -1];
            }
            if(k == 1){
                return min(nums1[index1],nums2[index2]);
            }

            int newIndex1 = min(index1+k/2-1,m-1);
            int newIndex2 = min(index2+k/2-1,n-1);

            if(nums1[newIndex1] >= nums2[newIndex2]){
                k -= newIndex2 - index2 +1;
                index2 = newIndex2+1;
            }else{
                k -= newIndex1 - index1 +1;
                index1 = newIndex1+1;
            }

        }
    }
};

File: c++/leetcode/array/Median_of_Two_Sorted_Arrays.go
package main

import (
	"fmt"
	"math"
)

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	k := int(math.Ceil((float64(len(nums1)) + float64(len(nums2))) / 2))

	if (len(nums1)+len(nums2))%2 == 0 {
		foo1 := getKthElement(nums1, nums2, k)
		foo2 := getKthElement(nums1, nums2, k+1)
		return float64(foo1+foo2) / 2
	} else {
		return float64(getKthElement(nums1, nums2, k))
	}

}

func getKthElement(nums1 []int, nums2 []int, k int) int {
	if len(nums1) == 0 {
		return nums2[k-1]
	}

	if len(nums2) == 0 {
		return nums1[k-1]
	}

	compareIdx := k / 2

	if compareIdx == 0 {
		return min(nums1[0], nums2[0])
	}

	nums1Idx := min(len(nums1)-1, compareIdx-1)
	nums2Idx := min(len(nums2)-1, compareIdx-1)

	if nums1[nums1Idx] >= nums2[nums2Idx] {
		if len(nums2) <= compareIdx {
			return getKthElement(nums1, []int{}, k-(nums2Idx+1))
		}
		return getKthElement(nums1, nums2[compareIdx:], k-(nums2Idx+1))
	} else {
		if len(nums1) <= compareIdx {
			return getKthElement([]int{}, nums2, k-(nums1Idx+1))
		}
		return getKthElement(nums1[compareIdx:], nums2, k-(nums1Idx+1))
	}
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func main() {
	nums1 := []int{1, 2}
	nums2 := []int{3, 4}
	fmt.Println(findMedianSortedArrays(nums1, nums2))
	//fmt.Println(getKthElement(nums1, nums2, 4))
}
